### DISCUSSION TITLE
Build system discussion

### DISCUSSION THREAD

**Discussion by Rot127** [aquynh/capstone#2290] [General]
We mentioned here and there that the `Makefile` build is considered deprecated and should be removed. Though it was never really discussed or tracked somewhere.

Supporting only `cmake` is indeed maybe not enough(?). We could add support for other build systems. Meson, Bazel?

Personally I am still in favor of deprecating the `Makefile` way. Its pretty messy IMHO.

Thoughts, objections etc. are very welcome.

**Comment by sevan** [comment]
There's a shift of moving to new build infrastructure, away from the traditional Makefile or autotools infra, whether it's to get away from perceived arcaneness or lack of performance. The thing is that the usual pick of the mix build tools that are currently in flavour have much higher dependency requirement than the code base they're actually used to build. I am capable of build this project with a C compiler which supports C99 + a copy of make. As an example for cmake and such I would need a much newer compiler (with C++11 support) + its dependencies. You could make the argument that I've gone off into the weeds and a current system not only would have the language support but would have everything ready to set off and that's true however such tooling imposes that only current mainstream systems are used because anything fringe will be very hard or impossible.
I would suggest the support of both legacy make infra is maintained as a fall through for the fringe, armed with a C99 compiler & make to be able to build capstone alongside modern build infra. That should help also clarify what is required to build the project.
If you need a hand, I'm happy to add autotools support as well.

**Comment by XVilka** [comment]
Makefile in its current form is undoubtedly a pain to maintain and improve. I think having just CMake in the near future is enough. Supporting both Meson and CMake at the same time could be problematic. As for Bazel - I don't see many requests for it just yet.

**Comment by trufae** [comment]
The cmake is imho way more unreadable than the makefile, and this makefile is indeed full of spaguetti and can be trimmed down to at least half of the lines. Considering make is the standard for buliding in UNIX systems I would prefer to keep it there.

I am not a big fan of CMake, as a packager and developer it has always been a source of problems, but it's there, it works and it's maintained, so I won't complain about removing it.

But I would like to introduce meson to the party. Many projects use capstone and build it with meson by maintaining the files on each project. It will be good if capstone could be distributed as a dependency for meson too.
